ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala

```diff
@@ -98,37 +78,48 @@ private[apiserver] final class ApiPartyManagementService private (
-            .allocateParty(party, displayName, submissionId))
-          .flatMap {
-            case SubmissionResult.Acknowledged =>
-              pollUntilPersisted(submissionId, ledgerEndBeforeRequest).flatMap {
-                case domain.PartyEntry.AllocationAccepted(_, partyDetails) =>
-                  Future.successful(
-                    AllocatePartyResponse(
-                      Some(
-                        PartyDetails(
-                          partyDetails.party,
-                          partyDetails.displayName.getOrElse(""),
-                          partyDetails.isLocal,
-                        ))))
-                case domain.PartyEntry.AllocationRejected(_, reason) =>
-                  Future.failed(ErrorFactories.invalidArgument(reason))
-              }(DE)
-            case r @ SubmissionResult.Overloaded =>
-              Future.failed(ErrorFactories.resourceExhausted(r.description))
-            case r @ SubmissionResult.InternalError(_) =>
-              Future.failed(ErrorFactories.internal(r.reason))
-            case r @ SubmissionResult.NotSupported =>
-              Future.failed(ErrorFactories.unimplemented(r.description))
-          }(DE)
-      }(DE)
-      .andThen(logger.logErrorsOnCall[AllocatePartyResponse])(DE)
+    // Execute subsequent transforms in the thread of the previous operation.
+    implicit val executionContext: ExecutionContext = DE
+
+    val response = for {
+      ledgerEndBeforeRequest <- transactionService.currentLedgerEnd()
+      submissionResult <- writeService.allocateParty(party, displayName, submissionId).toScala
+      entry <- submissionResult match {
+        case SubmissionResult.Acknowledged =>
+          SynchronousResponse.pollUntilPersisted(
+            partyManagementService
+              .partyEntries(ledgerEndBeforeRequest)
+              .collect {
+                case entry @ AllocationAccepted(Some(`submissionId`), _) => entry
+                case entry @ AllocationRejected(`submissionId`, _) => entry
+              },
+            30.seconds,
+          )(materializer)
+        case r @ SubmissionResult.Overloaded =>
+          Future.failed(ErrorFactories.resourceExhausted(r.description))
+        case r @ SubmissionResult.InternalError(_) =>
+          Future.failed(ErrorFactories.internal(r.reason))
+        case r @ SubmissionResult.NotSupported =>
+          Future.failed(ErrorFactories.unimplemented(r.description))
+      }
+      response <- entry match {
+        case domain.PartyEntry.AllocationAccepted(_, partyDetails) =>
+          Future.successful(
+            AllocatePartyResponse(
+              Some(
+                PartyDetails(
+                  partyDetails.party,
+                  partyDetails.displayName.getOrElse(""),
+                  partyDetails.isLocal,
+                ))))
+        case domain.PartyEntry.AllocationRejected(_, reason) =>
+          Future.failed(ErrorFactories.invalidArgument(reason))
+      }
```

Comment:
If you want to spend 10 minutes breaking this up in named methods the readability would probably gain much more than from removing the explicit `flatMap`.

Reply:
I think that's a solid idea. I did try that but abandoned it half-way because it wasn't looking amazing, but that was before some other changes that I think will make it easier.

Reply:
So… I moved a lot more behavior into `SynchronousResponse`. I hope I didn't go overboard.